zhusuan/variational/base.py

```diff
@@ -49,16 +49,24 @@ def __init__(self, log_joint, observed, latent):
         self._latent_logpdfs = dict(
             zip(latent_k, map(lambda x: x[1], latent_v)))
         self._joint_obs = merge_dicts(observed, self._latent_outputs)
+        self._dict_key = (log_joint,
+                          frozenset(latent_k),
+                          frozenset(map(tuple, latent_v)),
+                          frozenset(six.iteritems(observed)))
+
+    # Cache for log_p and log_q
+    _log_p, _log_q = {}, {}
 
     def _log_joint_term(self):
-        if not hasattr(self, '_log_p'):
-            self._log_p = self._log_joint(self._joint_obs)
-        return self._log_p
+        if self._dict_key not in self._log_p:
```

Comment:
if the cache is in class, should it still be accessed by self? I guess you can check the Context class that BayesianNet is inherited from. maybe some classmethod like _get_cache(), _set_cache() is good? And call them by type(self)._set_cache() or things like this.

Reply:
maybe just use classmethod to get _log_p is better? like the last commit?